# Introduction: The Power of Simple Questions

Imagine you are a doctor diagnosing a patient. You don't run every test simultaneously. Instead, you ask a sequence of questions: "Does the patient have a fever?" If yes, you ask, "Is there a rash?" If no, you might ask, "Is the pain localized?" This hierarchical process of asking questions to narrow down possibilities is the very essence of a Decision Tree.

A Decision Tree is a **non-parametric** (**no** assumptions about data distribution or functional form, allowing them to learn patterns directly from the data), **supervised learning** algorithm used for both classification and regression tasks. Its model forms a tree-like structure, mimicking human decision-making processes.



> **Parameteric vs Non-Parametric**: A Decision Tree (DT) is non-parametric, meaning it makes no assumptions about data distribution or functional form and has no fixed number of parameters—its complexity (splits, depth, etc.) adapts to the data; unlike parametric models (e.g., linear or logistic regression) that assume a specific functional form and have fixed parameters.


They **mimic human decision-making** by breaking down complex decisions into a series of simpler questions. A Decision Tree represents decisions in a flowchart-like, hierarchical structure that models possible consequences—including outcomes, costs, and probabilities and is highly interpretable.

<center>
<img src="https://media.geeksforgeeks.org/wp-content/uploads/20221212133528/dcsion.png" alt="Pandas Illustration" width="600">
</center>

They are especially popular because they require minimal data preprocessing, can handle both numerical and categorical data, and provide models that are easily interpretable.


It has many strengths:

* **Interpretability:** The resulting model is a white box. The rules for making a prediction can be easily understood and explained, even to non-experts.

* **Few Data Preprocessing Requirements:** They require little data preparation (e.g., no need for feature scaling or normalization).

* **Handles Mixed Data Types:** They can work with both numerical and categorical data.

* **Non-Linearity:** They can capture complex non-linear relationships between features and the target variable.


However, this simplicity can also be a weakness because it often leads to **overfitting**. When a tree grows too deep, it starts to learn the noise in the training data, not just the underlying pattern, which causes poor performance on new data.

In this chapter, we will break down how the **Decision Tree algorithm works**, understand how to create **good splitting questions**, learn how to **prune** an overly complex tree, and see how Decision Trees serve as the foundation for advanced ensemble methods like Random Forests and Gradient Boosting Machines

# Section 1: Basic Concepts of Decision Tree

Next, we will learn some basic concepts related to decision tree.

## Structure of a Decision Tree and Terminology

To understand how a tree is built, we must first understand its components. A Decision Tree is a graph composed of:

<center>
<img src="https://365datascience.com/resources/blog/rr6cuudl59r-decision-trees-image1.png" alt="Pandas Illustration" width="600">
</center>


* **Root Node:** The topmost node, representing the entire dataset. It is the starting point for the splitting process.

* **Internal Nodes (Decision Nodes):** Nodes that represent a decision point or a test on a specific feature. Each internal node splits the data into two or more branches.

* **Branches (Edges):** The outcome of a decision node test, leading to the next node.

* **Leaf Nodes (Terminal Nodes):** The final nodes in the tree that do not split further. Each leaf node represents a class label (in classification) or a continuous value (in regression) which is the final prediction.

The path from the root to a leaf represents a series of conjunctions (AND operations) of the conditions at the internal nodes, forming a classification rule.

### More Terminology

* **Sub-Tree:** A portion of the main tree.

* **Splitting:** The process of dividing a node into two or more sub-nodes.

* **Pruning:** The process of removing branches to reduce the complexity of the tree and prevent overfitting (we will talk about it later in this chapter).


### **Example**

Consider the classic “Play Tennis” dataset, where we want to predict whether to play tennis based on weather conditions (Outlook, Temperature, Humidity, Windy).

A simplified decision tree might look like:


<center>
<img src="https://spotintelligence.com/wp-content/uploads/2024/05/decision-tree-example.jpg" alt="Pandas Illustration" width="600">
</center>




# Section 2: The Tree-Building Algorithm: A Greedy, Recursive Process

The construction of a Decision Tree is a **greedy**, **recursive**, and **top-down** process, known as Recursive Binary Splitting (or partitioning).


> **Key Idea:** The core idea behind building a decision tree is **recursive partitioning**; repeatedly splitting the dataset based on an attribute that maximizes separation between classes (or minimizes error in regression).



## Detailed Concept of Decision Tree Learning Algorithm:

* **Greedy:** At each node, the algorithm selects the locally optimal split without considering the overall tree structure or future splits. It makes the best decision it can at that moment.

* **Recursive:** The process is repeated for each subset of the data created by previous splits.

* **Top-Down:** It begins at the root and successively splits the dataset into smaller groups.


### The Core Algorithm (**ID3** and its successors **C4.5**, **CART**):

The **Iterative Dichotomiser 3 (ID3)** algorithm is a decision tree induction algorithm that constructs decision trees using a top-down approach. The main goal of ID3 is to find the most informative attributes at each step to create partitions that yield the **highest information gain** or **decrease in impurity**.

Let's break down the ID3 algorithm into a step-by-step process:

- ➤ 1. **Start** at the root node with the entire dataset.

- ➤ 2. **For each feature in the dataset**, calculate a *"goodness" metric* (e.g., Information Gain, Gini Impurity) for all possible split points.

> In this chapter, we will focus on "Entropy" and "Information Gain".

- ➤ 3. Select the feature and split point that **maximizes this "goodness" metric**. This becomes the splitting rule for the node.

- ➤ 4. **Split the dataset at the current node** into child nodes based on the selected rule.

The process recursively **repeat steps 2-4**S for each child node until **a stopping criterion** is met.

**Common stopping criteria** include:

- The node is **"pure"** (homogeneous: all samples belong to the same class).
- No more attributes remain for further splitting.
- A stopping condition (like a predefined maximum depth) is reached.
- The node contains fewer samples than the predefined minimum threshold.
- A split does not produce a significant improvement (minimum gain threshold).

# Splitting Criteria (Attribute Selection Measures):

The algorithm needs a metric to determine the **"best" split**. For classification trees, the goal is to **maximize the homogeneity** of the resulting subsets. The main metrics are **Entropy** and **Information Gain** (Used in ID3, C4.5).

## Entropy

Entropy measures the randomness or **impurity** in the dataset.



> In information theory, the entropy of a random variable is the average level of "information", "surprise", or "uncertainty" inherent to the variable's possible outcomes.

In a decision tree, a node that contains a mix of different outcomes (for example, 2 Pass and 2 Fail) has **higher entropy**, meaning more disorder or uncertainty. If a node has only one type of outcome (for example, all Pass or all Fail), it is **pure and has low entropy**.

- **Maximum entropy = 1** → occurs when the outcomes are equally mixed (e.g., 50% Pass, 50% Fail).

- **Minimum entropy = 0** → occurs when all outcomes are the same (e.g., 100% Pass or 100% Fail).

> In short: Mixed outcomes → high entropy (uncertain) and Single outcome → low entropy (certain)


<!-- <center>
<img src="https://miro.medium.com/v2/resize:fit:1358/format:webp/1*E2qTQ1NSzy2Zb6YC8oqDvA.png" alt="Pandas Illustration" width="600">
</center> -->

<center>
<img src="https://miro.medium.com/v2/resize:fit:1400/1*ODTVm0g4PkSZnDMW91y_Qw.png" alt="Pandas Illustration" width="600">
</center>

The first thing to understand about Decision Trees is that they split the data **based on the input features into smaller groups** that are **more homogeneous (or “pure”)** with respect to the target variable.

For example, if the target variable has two possible outcomes — say 1 and 0 (represented by green and red dots) — the Decision Tree looks for splits in the features that create groups mostly containing either 1’s or 0’s, rather than a mix of both.

>In simple terms, the tree keeps splitting the data using feature values so that each group becomes as pure as possible in terms of the target outcome.

<center>
<img src="https://towardsdatascience.com/wp-content/uploads/2022/11/1HQsjuYNRaphQ0SFXnedqRA.png" alt="Pandas Illustration" width="600">
</center>

### **Definition**

Entropy is defined (or calculated) as:

$$
Entropy(S) = - \sum_{k=1}^{c} p_k \log_2(p_k)
$$

where:  
- $S$ = the current node (or dataset)  
- $c$ = number of classes  
- $p_k$ = proportion of data points in class $k$ within node $S$  


### **Properties of Entropy**

- **Entropy = 0:**  
  The node is **pure** — all samples belong to a single class (no uncertainty).  

- **Entropy = 1:**  
  The node is **maximally impure** — classes are **equally mixed** (e.g., 50% Pass, 50% Fail).  

- **Range:**  
  $ 0 \leq Entropy(S) \leq 1 $ for binary classification.  
  (For multi-class problems, entropy can be higher, depending on the number of classes.)  

---

### **Example**

If a node contains 4 samples:  
- 3 are **Pass**,  
- 1 is **Fail**,  

then:

$$
Entropy(S) = -\left( \frac{3}{4} \log_2 \frac{3}{4} + \frac{1}{4} \log_2 \frac{1}{4} \right)
$$

$$
Entropy(S) = -(0.75 \times -0.415) - (0.25 \times -2)
$$

$$
Entropy(S) \approx 0.81
$$

So, this node has some impurity but is not completely mixed.

### **Try it Yourself**

If a node contains **5 examples**, where **4 are Pass** and **1 is Fail**, calculate the entropy.

<!-- $$
Entropy(S) = -\left( \frac{4}{5} \log_2 \frac{4}{5} + \frac{1}{5} \log_2 \frac{1}{5} \right)
$$ -->

You can use the Python function below to verify your answer!

In [ ]:
import math

def entropy(class_counts):
    """
    Compute entropy for a list of class counts.
    class_counts: list of counts for each class
    """
    total = sum(class_counts)
    entropy_value = 0
    for count in class_counts:
        if count == 0:  # avoid log(0)
            continue
        p = count / total
        entropy_value -= p * math.log2(p)
    return entropy_value

# Example 1: 4 Pass, 1 Fail
print("Entropy (4 Pass, 1 Fail):", round(entropy([4, 1]), 3))

# Example 2: 2 Pass, 2 Fail (max impurity)
print("Entropy (2 Pass, 2 Fail):", round(entropy([2, 2]), 3))

# Example 3: All Pass (pure node)
print("Entropy (5 Pass, 0 Fail):", round(entropy([5, 0]), 3))

Entropy (4 Pass, 1 Fail): 0.722
Entropy (2 Pass, 2 Fail): 1.0
Entropy (5 Pass, 0 Fail): 0.0


Once we know how to calculate **Entropy**, we can measure how much **uncertainty** (or impurity) is reduced when we split a dataset using a feature. This reduction is called **Information Gain**.

## Information Gain (IG)
Information Gain measures the **reduction in entropy** achieved by a split. It tells us **how much uncertainty (impurity) is removed** by a particular split.  

### **Definition**

Formally, suppose a split divides the parent node $S$ into $V$ subsets:  
$$
S_1, S_2, ..., S_V
$$

Let:  
- $|S|$ = total number of samples in the parent node  
- $|S_v|$ = number of samples in child node $S_v$  

Information Gain for a feature **A** is defined as:

$$
IG(S, A) = Entropy(S) - \sum_{v \in Values(A)} \frac{|S_v|}{|S|} \, Entropy(S_v)
$$

where:  
- $S$ = parent node (original dataset)  
- $A$ = feature used for splitting  
- $S_v$ = subset of $S$ corresponding to value $v$ of feature $A$  
- $|S|$, $|S_v|$ = number of samples in $S$ and $S_v$, respectively  

The algorithm selects the **feature with the highest Information Gain** for splitting.

---

### **Intuition**

Information Gain measures how much a split **reduces impurity**:
- A **high IG** means the feature created **purer subsets** (good split).  
- A **low IG** means the feature did not reduce impurity much (poor split).  

Decision Trees choose the feature with the **highest Information Gain** at each step.

---

### **Example**

Suppose a dataset has 10 samples:  
- 6 **Pass**, 4 **Fail**  

Then the initial (parent) entropy is:

$$
Entropy(S) = -\left( \frac{6}{10}\log_2\frac{6}{10} + \frac{4}{10}\log_2\frac{4}{10} \right) \approx 0.971
$$

Now we split based on **Feature A** (e.g., “Study Hours”) which has two possible values:  
- **A = High:** 5 samples → 4 Pass, 1 Fail  
- **A = Low:** 5 samples → 2 Pass, 3 Fail  

Compute entropy for each group:

$$
Entropy(S_{High}) = -\left( \frac{4}{5}\log_2\frac{4}{5} + \frac{1}{5}\log_2\frac{1}{5} \right) \approx 0.722
$$

$$
Entropy(S_{Low}) = -\left( \frac{2}{5}\log_2\frac{2}{5} + \frac{3}{5}\log_2\frac{3}{5} \right) \approx 0.971
$$

Then the weighted average entropy after splitting is:

$$
Entropy_{after} = \frac{5}{10}(0.722) + \frac{5}{10}(0.971) = 0.847
$$

Finally, the **Information Gain** is:


<center>
<img src="https://miro.medium.com/v2/resize:fit:1400/format:webp/1*KjVvgo1Ks0WJb6i3tHwdBA.png" alt="Pandas Illustration" width="600">
</center>


$$
IG(S, A) = 0.971 - 0.847 = 0.124
$$

So, splitting on **Feature A** reduces uncertainty by **0.124 bits**.

### **Additional Example:**
Example illustrating the calculation of information gain. Source: Hendler 2018, slide 46.

<center>
<img src="https://devopedia.org/images/article/168/4250.1555312917.jpg" alt="Pandas Illustration" width="600">
</center>

---

### **Try it Yourself**

If you have another feature **B** that splits the same data into:
- **B = Yes:** 6 samples (5 Pass, 1 Fail)  
- **B = No:** 4 samples (1 Pass, 3 Fail)

Try calculating:

$$
Entropy(S_{Yes}), \; Entropy(S_{No}), \; \text{and} \; IG(S, B)
$$

Which feature gives a higher Information Gain — **A** or **B**?


In [ ]:
def information_gain(parent_counts, split_counts_list):
    """
    Compute Information Gain.
    parent_counts: list of counts in the parent node
    split_counts_list: list of lists, each sublist contains counts in a child node
    """
    total_parent = sum(parent_counts)
    parent_entropy = entropy(parent_counts)

    weighted_entropy = 0
    for counts in split_counts_list:
        weight = sum(counts) / total_parent
        weighted_entropy += weight * entropy(counts)

    ig = parent_entropy - weighted_entropy
    return ig

# Example: Feature A split
# Parent node: 6 Pass, 4 Fail
parent = [6, 4]

# Feature A splits:
# High: 4 Pass, 1 Fail
# Low: 2 Pass, 3 Fail
split_A = [[4, 1], [2, 3]]

print("Information Gain for Feature A:", round(information_gain(parent, split_A), 3))

# Feature B splits:
# Yes: 5 Pass, 1 Fail
# No: 1 Pass, 3 Fail
split_B = [[5, 1], [1, 3]]

print("Information Gain for Feature B:", round(information_gain(parent, split_B), 3))

Information Gain for Feature A: 0.125
Information Gain for Feature B: 0.256


# How Decision Trees Decide Which Feature to Split (Using Entropy and Information Gain)

Decision Trees use **Entropy** and **Information Gain** to decide **how to split the data at each node**. The goal is to create **subgroups (child nodes) that are as pure as possible** with respect to the target variable.

> The core idea of a Decision Tree is **recursive splitting**: at each node, we pick the **feature that best separates the data** into purer subgroups.

---

### **Step-by-Step Process**

1. **Start at the root node**  
   - The root node contains the entire dataset.  
   - Calculate the **Entropy** of the root node to measure the overall uncertainty.

<!-- 2. **Evaluate all features for splitting**  
   - For each feature, simulate splitting the data according to its possible values.  
   - Compute the **weighted entropy** of the resulting child nodes.  
   - Calculate the **Information Gain** for that feature:  
$IG(S, A) = Entropy(S) - \sum_{v \in Values(A)} \frac{|S_v|}{|S|} \, Entropy(S_v)$ -->

2. **Evaluate each feature**  
   - For each feature, split the dataset according to its possible values.  
   - Compute the **entropy of each child node** and the **weighted average entropy**:
$$
Entropy_{after\_split} = \sum_{v \in Values(A)} \frac{|S_v|}{|S|} \, Entropy(S_v)
$$

   - Calculate **Information Gain**:

$$
IG(S, A) = Entropy(S) - Entropy_{after\_split}
$$

   - A **higher IG** means the feature produces **purer child nodes**.

3. **Choose the best feature to split**  
   - Compare the Information Gain of all features.  
   - Select the **feature with the highest IG** for splitting.  
   - This ensures each split reduces the most uncertainty.
      - Creates child nodes that are **more homogeneous** than the parent


<!-- 3. **Choose the best feature to split**  
   - The feature with the **highest Information Gain** is selected to split the current node.  
   - This creates child nodes that are **more homogeneous** than the parent. -->


4. **Repeat recursively for each child node**  
   - Treat each child node as a new node and repeat Steps 1–3.  
   - Continue until:
     - Nodes are **pure** (entropy = 0), or  
     - A **stopping condition** is reached (e.g., max depth, min samples per node).


<!-- 4. **Repeat recursively for each child node**  
   - For each child node, repeat the process: calculate entropy, evaluate features, and split using the feature with highest information gain.  
   - Continue until:  
     - Nodes are **pure** (all samples in the node belong to the same class), or  
     - A **stopping criterion** is reached (e.g., max depth, min samples per node). -->

---

### **Key Idea**

- **Entropy** tells us how **mixed** the node is.  
- **Information Gain** tells us how **good a feature is at reducing uncertainty**.  
- Decision Trees **always pick the feature that maximizes Information Gain** to make the next split.  

---

## Detailed Example 01: How Decision Trees Decide Which Feature to Split

Consider an example where we are building a decision tree to predict whether a loan given to a person would result in a write-off or not. Our entire population consists of 30 instances. 16 belong to the write-off class and the other 14 belong to the non-write-off class. We have two features, namely “Balance” that can take on two values -> “< 50K” or “>50K” and “Residence” that can take on three values -> “OWN”, “RENT” or “OTHER”.


# Overfitting and Pruning

## Overfitting in Decision Tree
Decision trees are prone to overfitting, meaning they become too complex and capture noise in the training data (**perfectly fitting the training data**), but performing poorly on unseen data (**poor generalization**). This often happens with very deep trees with many nodes that achieve perfect classification on the training set (pure leaf nodes).

## Pruning Methods

**Pruning** is the technique used to combat overfitting by reducing the size of the tree. There are two main strategies:


1.   **Pre-pruning (Early Stopping)** : Stop growing the tree early before it becomes too complex. How? by setting parameters like:

*  Maximum depth of the tree.
*  Minimum number of samples required to split a node.
*  Minimum number of samples required to be in a leaf node.

2.  **Post-pruning:** Grow a full tree and then remove branches that do not improve accuracy on validation data. The most common technique is Reduced Error Pruning.
